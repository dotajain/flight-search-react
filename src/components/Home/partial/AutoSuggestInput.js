import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import Autosuggest from 'react-autosuggest';

import { searchCities } from '../../../actions';

class AutoSuggestInput extends Component {
  constructor() {
    super();
    this.state = {
      value: '',
      suggestions: [],
      cities: [],
    };
  }

  onChange = (event, { newValue }) => {
    this.setState({ value: newValue }, () => {
      if (this.state.value.length >= 2) {
        this.props.searchCities();
      }
    });
  };

  getSuggestions = value => {
    const inputValue = value.trim().toLowerCase();
    const inputLength = inputValue.length;
    return inputLength === 0
      ? []
      : this.state.cities.filter(
          city =>
            city.cityName.toLowerCase().slice(0, inputLength) === inputValue,
        );
  };

  onSuggestionsFetchRequested = ({ value }) => {
    if (this.state.cities) {
      this.setState({ suggestions: this.getSuggestions(value) });
    }
  };

  onSuggestionsClearRequested = () => {
    this.setState({ suggestions: [] });
  };

  componentWillReceiveProps(nextProps) {
    if (nextProps.flightSearch.cities) {
      this.setState({ cities: nextProps.flightSearch.cities });
    }
  }

  renderSuggestion = suggestion => (
    <div className="react-autosuggest__result">
      <div className="airport-code">{suggestion.airportCode}</div>
      <div className="text-cntr">
        <div className="city">{`${suggestion.cityName}, ${suggestion.countryName}`}</div>
        <div className="airport">{suggestion.airportName}</div>
      </div>
    </div>
  );

  render() {
    const { value, suggestions } = this.state;
    const { placeholder, id, name, label } = this.props;

    const inputProps = {
      placeholder,
      value,
      onChange: this.onChange,
      id,
      name,
    };

    const getSuggestionValue = suggestion =>
      `${suggestion.airportCode} - ${suggestion.cityName}`;

    return (
      <div className="form__field form__field--text">
        <label htmlFor={name} className="form__label">
          {label}
        </label>
        <div className="form__field--contents">
          <Autosuggest
            suggestions={suggestions}
            onSuggestionsFetchRequested={this.onSuggestionsFetchRequested}
            onSuggestionsClearRequested={this.onSuggestionsClearRequested}
            getSuggestionValue={getSuggestionValue}
            renderSuggestion={this.renderSuggestion}
            inputProps={inputProps}
          />
        </div>
      </div>
    );
  }
}

AutoSuggestInput.propTypes = {
  id: PropTypes.string,
  name: PropTypes.string,
  searchCities: PropTypes.func,
  flightSearch: PropTypes.object,
  placeholder: PropTypes.string,
  label: PropTypes.string,
};

const mapStateToProps = state => ({ flightSearch: state.flightSearch });
export default connect(mapStateToProps, { searchCities })(AutoSuggestInput);
