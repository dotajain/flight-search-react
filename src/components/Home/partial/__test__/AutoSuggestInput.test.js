import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { store } from '../../../../store';

import AutoSuggestInput from '../AutoSuggestInput';

global.fetch = jest.fn(
  () =>
    new Promise(resolve => {
      process.nextTick(() => resolve({ json: () => ({}) }));
    }),
);

it('Properly rendered AutoSuggestInput Component', () => {
  const component = renderer.create(<Provider store={store}><AutoSuggestInput /></Provider>);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
