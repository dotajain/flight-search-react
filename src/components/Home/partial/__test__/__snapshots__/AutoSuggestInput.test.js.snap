// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Properly rendered AutoSuggestInput Component 1`] = `
<div
  className="form__field form__field--text"
>
  <label
    className="form__label"
    htmlFor={undefined}
  />
  <div
    className="form__field--contents"
  >
    <div
      className="react-autosuggest__container"
    >
      <input
        aria-activedescendant={null}
        aria-autocomplete="list"
        aria-expanded={false}
        aria-haspopup={false}
        aria-owns="react-autowhatever-1"
        autoComplete="off"
        className="react-autosuggest__input"
        id={undefined}
        name={undefined}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        placeholder={undefined}
        role="combobox"
        type="text"
        value=""
      />
      <div
        className="react-autosuggest__suggestions-container"
        id="react-autowhatever-1"
      />
    </div>
  </div>
</div>
`;
